main/util/src/mill/util/CoursierSupport.scala

```diff
@@ -208,60 +227,126 @@ trait CoursierSupport {
       coursierCacheCustomizer: Option[
         coursier.cache.FileCache[Task] => coursier.cache.FileCache[Task]
-      ] = None
+      ] = None,
+      sameDependencyVersions: Seq[Set[(String, String)]] = Seq()
   ): (Seq[Dependency], Resolution) = {
 
-    val cachePolicies = coursier.cache.CacheDefaults.cachePolicies
+    val rootDeps = deps.iterator
+      .map(d => mapDependencies.fold(d)(_.apply(d)))
+      .toSeq
 
-    val forceVersions = force.iterator
-      .map(mapDependencies.getOrElse(identity[Dependency](_)))
+    val forceVersions: Map[Module, String] = force.iterator
+      .map(d => mapDependencies.fold(d)(_.apply(d)))
       .map { d => d.module -> d.version }
       .toMap
 
-    val start0 = Resolution()
-      .withRootDependencies(
-        deps.iterator.map(mapDependencies.getOrElse(identity[Dependency](_))).toSeq
-      )
-      .withForceVersions(forceVersions)
-      .withMapDependencies(mapDependencies)
-
-    val start = customizer.getOrElse(identity[Resolution](_)).apply(start0)
-
-    val resolutionLogger = ctx.map(c => new TickerResolutionLogger(c))
-    val coursierCache0 = resolutionLogger match {
-      case None => coursier.cache.FileCache[Task]().withCachePolicies(cachePolicies)
-      case Some(l) =>
-        coursier.cache.FileCache[Task]()
-          .withCachePolicies(cachePolicies)
-          .withLogger(l)
+    val coursierCache: FileCache[Task] = FileCache[Task]()
+      .noCredentials
+      .pipe { cache =>
+        coursierCacheCustomizer.fold(cache)(c => c.apply(cache))
+      }.pipe { cache =>
+        ctx.fold(cache)(c => cache.withLogger(new TickerResolutionLogger(c)))
+      }
+
+    val rules = sameDependencyVersions.map { same =>
+      val modMatchers = same.map { case (org, name) =>
+        ModuleMatcher(Organization(org), ModuleName(name))
+      }
+      (SameVersion(modMatchers), RuleResolution.TryResolve)
     }
-    val coursierCache = coursierCacheCustomizer.getOrElse(
-      identity[coursier.cache.FileCache[Task]](_)
-    ).apply(coursierCache0)
 
-    val fetches = coursierCache.fetchs
+    {
+      val resolutionParams = ResolutionParams()
+        .withForceVersion(forceVersions)
+        .withRules(rules)
 
-    val fetch = coursier.core.ResolutionProcess.fetch(repositories, fetches.head, fetches.tail)
+      val resolve = Resolve()
+        .withCache(coursierCache)
+        .withDependencies(rootDeps)
+        .withRepositories(repositories)
+        .withResolutionParams(resolutionParams)
 
-    import scala.concurrent.ExecutionContext.Implicits.global
+      val resolution0 = retry(
+        ctx = ctx,
+        errorMsgExtractor = (r: Resolution) => r.errors.flatMap(_._2)
+      )(resolve.run)
 
-    val resolution =
-      retry(ctx = ctx, errorMsgExtractor = (r: Resolution) => r.errors.flatMap(_._2)) {
-        () => start.process.run(fetch).unsafeRun()
-      }
+      val resolution = resolution0
+        .withMapDependencies(mapDependencies)
```

Comment:
I think this `mapDependencies` is applied too late. It must be applied before resolution instead. But I can't find any other way in the new `Resolve` API to pass a map function. @alexarchambault Do you have an idea?

Reply:
Related?

* https://github.com/coursier/coursier/pull/3072